### PR TITLE
Feature user application

### PR DESCRIPTION
**Pull request by Burito2021**

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -13,19 +13,35 @@ java {
...
+	testImplementation "org.junit.jupiter:junit-jupiter:${junit}"
 	testRuntimeOnly "org.junit.platform:junit-platform-launcher"
 	implementation "org.hibernate.validator:hibernate-validator:${hibernate}"
 	implementation "com.fasterxml.jackson.core:jackson-databind:${fasterxml}"
 	implementation "com.fasterxml.jackson.core:jackson-core"
 	implementation "com.fasterxml.jackson.core:jackson-annotations"
-	testImplementation "org.hamcrest:hamcrest:${hamcrest}"
+	implementation "org.springframework.boot:spring-boot-starter-data-jpa"
+	implementation "org.postgresql:postgresql:${postgresql}"
+	implementation "org.testcontainers:testcontainers:${testcontainers}"
```
Comment (MaximDzhezhelo) [inline build.gradle:41]: are you sure that testcontainers should be in general dependencies?

File: src/main/java/net/library/controller/UserController.java [hunk 2]
```diff
@@ -25,14 +28,30 @@ public UserController(final UserService service) {
...
-        service.initUserLst();
+    @GetMapping()
+    public ResponseEntity<List<User>> getAllUsers() {
+
+        return new ResponseEntity<>(service.getAllUsers(),HttpStatus.OK);
     }
 
-    @GetMapping()
-    public ResponseEntity<List<User>> getUsers() {
+    @PostMapping("/create")
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/controller/UserController.java:37]: use please best practices for rest api naming (https://restfulapi.net/resource-naming/)

File: src/main/java/net/library/config/PostgresTestContainerConfig.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
+package net.library.config;
+
+import org.testcontainers.containers.PostgreSQLContainer;
+
+
+public class PostgresTestContainerConfig {
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/config/PostgresTestContainerConfig.java:6]: All test utils create and use in the tests

File: src/main/java/net/library/controller/UserController.java [hunk 4]
```diff
@@ -25,14 +28,30 @@ public UserController(final UserService service) {
         this.service = service;
     }
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/controller/UserController.java:30]: Remove extra/useless controller

File: src/main/java/net/library/controller/BookController.java [hunk 5]
```diff
@@ -45,7 +46,7 @@ public ResponseEntity<HttpCidSuccessResponse> addBook(@Valid @RequestBody Book b
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/controller/BookController.java:46]: remove useless logger

File: src/main/java/net/library/controller/BookController.java [hunk 6]
```diff
@@ -45,7 +46,7 @@ public ResponseEntity<HttpCidSuccessResponse> addBook(@Valid @RequestBody Book b
 
         service.addBook(book);
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/controller/BookController.java:47]: I don't see logic in the services. If you use db you should not use mock

File: src/main/java/net/library/model/dto/request/UserRequest.java [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+    @NotNull
+    @NotEmpty
+    private String name;
+
+    @NotNull
+    @NotEmpty
+    private String surname;
+
+    @NotEmpty
+    @NotEmpty
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/model/dto/request/UserRequest.java:21]: Why is the annotation found twice

File: src/main/java/net/library/model/dto/request/UsersTest.java [hunk 8]
```diff
@@ -1,11 +1,11 @@
-package net.library.object;
+package net.library.model.dto.request;
 
-public class User {
+public class UsersTest {
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/model/dto/request/UsersTest.java:3]: What is this, and why is this class with the test prefix lying in the general folder?

File: src/main/java/net/library/model/dto/response/HttpCidSuccessResponse.java [hunk 9]
```diff
@@ -1,4 +1,4 @@
-package net.library.exception;
+package net.library.model.dto.response;
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/model/dto/response/HttpCidSuccessResponse.java:1]: It is incorrect structure/architecture of your code. You should return the basic type of controller. For example for BookController you should return book. But Book dto may extend Success response for example. But make a better name that Success response

File: src/main/java/net/library/service/UserService.java [hunk 10]
```diff
@@ -1,39 +1,43 @@
...
-        users.clear();
-        LOGGER.info("All the users deleted");
+       return   StreamSupport.stream(userRepository.findAll().spliterator(),false).toList();
     }
 
-    public void addUser(User user){
+    public void saveUser(User user){
+
+        userRepository.save(user);
+        LOGGER.info("User saved");
```
Comment (MaximDzhezhelo) [inline src/main/java/net/library/service/UserService.java:37]: don't do it. if you need this info you should enable basic logging

File: src/main/resources/sql/schema_data.sql [hunk 11]
```diff
@@ -0,0 +1,24 @@
...
+    PHONE_NUMBER VARCHAR(15),
+    ADDRESS VARCHAR(300),
+    UPDATED_AT TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    CREATED_AT TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
+    DELETED_AT TIMESTAMP
+);
+
+
+--INSERT INTO USERS (USERNAME, NAME, SURNAME, EMAIL, PHONE_NUMBER, ADDRESS)
+--VALUES
```
Comment (MaximDzhezhelo) [inline src/main/resources/sql/schema_data.sql:19]: remove it. If you need an additional file for inserts you should create it

File: src/main/resources/init.sql [hunk 12]
```diff
@@ -0,0 +1,24 @@
...
+CREATE TABLE USERS
+(
+    ID UUID DEFAULT uuid_generate_v4() PRIMARY KEY,
+    USERNAME VARCHAR(200) UNIQUE NOT NULL,
+    NAME VARCHAR(200) NOT NULL,
+    SURNAME VARCHAR(200) NOT NULL,
+    EMAIL VARCHAR(100) NOT NULL,
+    PHONE_NUMBER VARCHAR(15),
+    ADDRESS VARCHAR(300),
+    UPDATED_AT TIMESTAMP DEFAULT CURRENT_TIMESTAMP,
```
Comment (MaximDzhezhelo) [inline src/main/resources/init.sql:12]: why do you have two different files with the same logic?

File: src/test/java/net/library/controller/BookControllerTest.java [hunk 13]
```diff
@@ -33,17 +39,27 @@ public BookControllerTest(BookService service) {
         this.service = service;
     }
 
+    @BeforeAll
+    static void setup() {
+        postgreSQLContainer = PostgresTestContainerConfig.getPostgreInstance();
```
Comment (MaximDzhezhelo) [inline src/test/java/net/library/controller/BookControllerTest.java:44]: you should init db and upload data before each start up in the test

File: src/test/java/net/library/controller/BookControllerTest.java [hunk 14]
```diff
@@ -33,17 +39,27 @@ public BookControllerTest(BookService service) {
...
+    }
```
Comment (MaximDzhezhelo) [inline src/test/java/net/library/controller/BookControllerTest.java:45]: use test resources for test purposes

**Review by MaximDzhezhelo — CHANGES_REQUESTED** [review]

File: src/test/java/net/library/controller/BookControllerTest.java (hunk 14 above)
Comment (Burito2021) [reply src/test/java/net/library/controller/BookControllerTest.java:45]: +

File: src/test/java/net/library/controller/BookControllerTest.java (hunk 13 above)
Comment (Burito2021) [reply src/test/java/net/library/controller/BookControllerTest.java:44]: +

File: src/main/resources/init.sql (hunk 12 above)
Comment (Burito2021) [reply src/main/resources/init.sql:12]: +

File: src/main/java/net/library/service/UserService.java (hunk 10 above)
Comment (Burito2021) [reply src/main/java/net/library/service/UserService.java:37]: I do not understand. When you say basic logging what do you mean?

File: src/main/resources/sql/schema_data.sql (hunk 11 above)
Comment (Burito2021) [reply src/main/resources/sql/schema_data.sql:19]: +

File: src/main/java/net/library/model/dto/request/UsersTest.java (hunk 8 above)
Comment (Burito2021) [reply src/main/java/net/library/model/dto/request/UsersTest.java:3]: +

File: src/main/java/net/library/model/dto/request/UserRequest.java (hunk 7 above)
Comment (Burito2021) [reply src/main/java/net/library/model/dto/request/UserRequest.java:21]: +

File: src/main/java/net/library/controller/BookController.java (hunk 6 above)
Comment (Burito2021) [reply src/main/java/net/library/controller/BookController.java:47]: +

File: src/main/java/net/library/controller/BookController.java (hunk 5 above)
Comment (Burito2021) [reply src/main/java/net/library/controller/BookController.java:46]: +

**Pull request closed without merging**